content/300-guides/150-testing/150-integration-testing.mdx

```diff
@@ -1,13 +1,458 @@
+
+<details><summary>Ecommerce schema</summary>
+
+```prisma file=schema.prisma
+// Can have 1 customer
+// Can have many order details
+model Order {
+  id           Int            @id @default(autoincrement())
+  createdAt    DateTime       @default(now())
+  customer     Customer       @relation(fields: [customerId], references: [id])
+  customerId   Int
+  orderDetails OrderDetails[]
+}
+
+// Can have 1 order
+// Can have many products
+model OrderDetails {
+  id        Int     @id @default(autoincrement())
+  products  Product @relation(fields: [productId], references: [id])
+  productId Int
+  order     Order   @relation(fields: [orderId], references: [id])
+  orderId   Int
+  total     Decimal
+  quantity  Int
+}
+
+// Can have many order details
+// Can have 1 category
+model Product {
+  id           Int            @id @default(autoincrement())
+  name         String
+  description  String
+  price        Decimal
+  sku          Int
+  orderDetails OrderDetails[]
+  category     Category       @relation(fields: [categoryId], references: [id])
+  categoryId   Int
+}
+
+// Can have many products
+model Category {
+  id       Int       @id @default(autoincrement())
+  name     String
+  products Product[]
+}
+
+// Can have many orders
+model Customer {
+  id      Int     @id @default(autoincrement())
+  email   String  @unique
+  address String?
+  name    String?
+  orders  Order[]
+}
+
+```
+
+</details>
+
+The guide uses a singleton pattern for Prisma Client setup. Refer to the [singleton](./unit-testing#singleton) docs for a walk through of how to set that up.
+
+## Add Docker to your project
+
+With Docker and Docker compose both installed on your machine you can use them in your project.
+
+1. Begin by creating a `docker-compose.yml` file at your projects root. Here you will add a Postgres image and specify the environments credentials.
+
+```yml file=docker-compose.yml
+# Set the version of docker compose to use
+version: '3.9'
+
+# The containers that compose the project
+services:
+  db:
+    image: postgres:13
+    restart: always
+    container_name: integration-tests-prisma
+    ports:
+      - '5433:5432'
+    environment:
+      POSTGRES_USER: prisma
+      POSTGRES_PASSWORD: prisma
+      POSTGRES_DB: tests
+```
+
+> **Note**: The compose version used here (`3.9`) is the latest at the time of writing, if you are following along be sure to use the same version for consistency.
+
+The `docker-compose.yml` file defines the following:
+
+- The Postgres image (`postgres`) and version tag (`:13`). This will be downloaded if you do not have it locally available.
+- The port `5433` is mapped to the internal (Postgres default) port `5432`. This will be the port number the database is exposed on externally.
```

Comment:
Why not mapping both to `5432`?

Reply:
I had issues with this and found that chaging the port solved it. I was going to document that but i cant find the reason why it didnt work. I though people might ask that exact question though....

Reply:
I think it only helps if you already have a PostgreSQL database service started locally, Then the 5432 port is already occupied and Docker cannot use it.

Reply:
Ohhh. Hmm. But i dont have anything else running on that port, i checked. Lemme double check again now and change port. It would make much more sense to map to the same port as thats what people expect too

Reply:
Yep, it doesnt work for me. I cant authenticate

Reply:
Docker compose logs probably show something about the port if it's already used / cannot be used

Reply:
Nope:

```
integration-tests-prisma | 2021-05-12T09:24:21.405053400Z The files belonging to this database system will be owned by user "postgres".
integration-tests-prisma | 2021-05-12T09:24:21.405094600Z This user must also own the server process.
integration-tests-prisma | 2021-05-12T09:24:21.405097800Z
integration-tests-prisma | 2021-05-12T09:24:21.405256600Z The database cluster will be initialized with locale "en_US.utf8".
integration-tests-prisma | 2021-05-12T09:24:21.405264900Z The default database encoding has accordingly been set to "UTF8".
integration-tests-prisma | 2021-05-12T09:24:21.405267700Z The default text search configuration will be set to "english".
integration-tests-prisma | 2021-05-12T09:24:21.405269900Z
integration-tests-prisma | 2021-05-12T09:24:21.405271900Z Data page checksums are disabled.
integration-tests-prisma | 2021-05-12T09:24:21.405273900Z
integration-tests-prisma | 2021-05-12T09:24:21.405370800Z fixing permissions on existing directory /var/lib/postgresql/data ... ok
integration-tests-prisma | 2021-05-12T09:24:21.406048100Z creating subdirectories ... ok
integration-tests-prisma | 2021-05-12T09:24:21.406186800Z selecting dynamic shared memory implementation ... posix
integration-tests-prisma | 2021-05-12T09:24:21.416945200Z selecting default max_connections ... 100
integration-tests-prisma | 2021-05-12T09:24:21.430756400Z selecting default shared_buffers ... 128MB
integration-tests-prisma | 2021-05-12T09:24:21.445033200Z selecting default time zone ... Etc/UTC
integration-tests-prisma | 2021-05-12T09:24:21.446501000Z creating configuration files ... ok
integration-tests-prisma | 2021-05-12T09:24:21.556875800Z running bootstrap script ... ok
integration-tests-prisma | 2021-05-12T09:24:21.893263400Z performing post-bootstrap initialization ... ok
integration-tests-prisma | 2021-05-12T09:24:22.511826500Z syncing data to disk ... ok
integration-tests-prisma | 2021-05-12T09:24:22.511876100Z
integration-tests-prisma | 2021-05-12T09:24:22.511880600Z
integration-tests-prisma | 2021-05-12T09:24:22.511883200Z Success. You can now start the database server using:
integration-tests-prisma | 2021-05-12T09:24:22.511885600Z
integration-tests-prisma | 2021-05-12T09:24:22.511887700Z     pg_ctl -D /var/lib/postgresql/data -l logfile start
integration-tests-prisma | 2021-05-12T09:24:22.511889800Z
integration-tests-prisma | 2021-05-12T09:24:22.511838500Z initdb: warning: enabling "trust" authentication for local connections
integration-tests-prisma | 2021-05-12T09:24:22.511898500Z You can change this by editing pg_hba.conf or using the option -A, or
integration-tests-prisma | 2021-05-12T09:24:22.511900800Z --auth-local and --auth-host, the next time you run initdb.
integration-tests-prisma | 2021-05-12T09:24:22.546438800Z waiting for server to start....2021-05-12 09:24:22.546 UTC [48] LOG:  starting PostgreSQL 13.2 (Debian 13.2-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
integration-tests-prisma | 2021-05-12T09:24:22.553264100Z 2021-05-12 09:24:22.553 UTC [48] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
integration-tests-prisma | 2021-05-12T09:24:22.574358000Z 2021-05-12 09:24:22.574 UTC [49] LOG:  database system was shut down at 2021-05-12 09:24:21 UTC
integration-tests-prisma | 2021-05-12T09:24:22.582641700Z 2021-05-12 09:24:22.582 UTC [48] LOG:  database system is ready to accept connections
integration-tests-prisma | 2021-05-12T09:24:22.620963500Z  done
integration-tests-prisma | 2021-05-12T09:24:22.620991400Z server started
integration-tests-prisma | 2021-05-12T09:24:22.869248400Z CREATE DATABASE
integration-tests-prisma | 2021-05-12T09:24:22.869755000Z 
integration-tests-prisma | 2021-05-12T09:24:22.869880300Z
integration-tests-prisma | 2021-05-12T09:24:22.869981000Z /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*
integration-tests-prisma | 2021-05-12T09:24:22.869985200Z
integration-tests-prisma | 2021-05-12T09:24:22.870780300Z 2021-05-12 09:24:22.870 UTC [48] LOG:  received fast shutdown request
integration-tests-prisma | 2021-05-12T09:24:22.877711700Z waiting for server to shut down....2021-05-12 09:24:22.877 UTC [48] LOG:  aborting any active transactions
integration-tests-prisma | 2021-05-12T09:24:22.878644700Z 2021-05-12 09:24:22.878 UTC [48] LOG:  background worker "logical replication launcher" (PID 55) exited with exit code 1
integration-tests-prisma | 2021-05-12T09:24:22.878755800Z 2021-05-12 09:24:22.878 UTC [50] LOG:  shutting down
integration-tests-prisma | 2021-05-12T09:24:22.920355100Z 2021-05-12 09:24:22.920 UTC [48] LOG:  database system is shut down
integration-tests-prisma | 2021-05-12T09:24:22.971438000Z  done
integration-tests-prisma | 2021-05-12T09:24:22.971467200Z server stopped
integration-tests-prisma | 2021-05-12T09:24:22.971752200Z
integration-tests-prisma | 2021-05-12T09:24:22.971780100Z PostgreSQL init process complete; ready for start up.
integration-tests-prisma | 2021-05-12T09:24:22.971782900Z
integration-tests-prisma | 2021-05-12T09:24:22.999460300Z 2021-05-12 09:24:22.999 UTC [1] LOG:  starting PostgreSQL 13.2 (Debian 13.2-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
integration-tests-prisma | 2021-05-12T09:24:22.999794900Z 2021-05-12 09:24:22.999 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
integration-tests-prisma | 2021-05-12T09:24:22.999817600Z 2021-05-12 09:24:22.999 UTC [1] LOG:  listening on IPv6 address "::", port 5432
integration-tests-prisma | 2021-05-12T09:24:23.014013600Z 2021-05-12 09:24:23.013 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
integration-tests-prisma | 2021-05-12T09:24:23.029548300Z 2021-05-12 09:24:23.029 UTC [76] LOG:  database system was shut down at 2021-05-12 09:24:22 UTC
integration-tests-prisma | 2021-05-12T09:24:23.039222500Z 2021-05-12 09:24:23.038 UTC [1] LOG:  database system is ready to accept connections
```

Reply:
Thats me starting it on 5433 as i cant start it on 5432